chia/data_layer/data_layer.py

```diff
@@ -136,3 +141,137 @@ async def get_root(self, store_id: bytes32) -> Optional[bytes32]:
             self.log.error(f"Failed to get root for {store_id.hex()}")
             return None
         return res.node_hash
+
+    async def _validate_batch(
+        self,
+        tree_id: bytes32,
+        to_check: List[SingletonRecord],
+        min_generation: int,
+        max_generation: int,
+    ) -> bool:
+        last_checked_hash: Optional[bytes32] = None
+        for record in to_check:
+            # Ignore two consecutive identical root hashes, as we've already validated it.
+            if last_checked_hash is not None and record.root == last_checked_hash:
+                continue
+            # Pick the latest root in our data store with the desired hash, before our already validated data.
+            root: Optional[Root] = await self.data_store.get_last_tree_root_by_hash(
+                tree_id, record.root, max_generation
+            )
+            if root is None or root.generation < min_generation:
+                return False
+
+            self.log.info(f"Validated chain hash {record.root} in downloaded datastore.")
+            max_generation = root.generation
+            last_checked_hash = record.root
+
+        return True
+
+    async def fetch_and_validate(self, subscription: Subscription) -> None:
+        tree_id = subscription.tree_id
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id)
+        if singleton_record is None:
+            return
+        if singleton_record.generation == uint32(0):
+            return
+        old_root: Optional[Root] = None
+        if await self.data_store.tree_id_exists(tree_id=tree_id):
+            old_root = await self.data_store.get_tree_root(tree_id=tree_id)
+        wallet_current_generation = await self.data_store.get_wallet_generation(tree_id)
+        assert int(wallet_current_generation) <= singleton_record.generation
+        # Wallet generation didn't change, so no new data committed on chain.
+        if wallet_current_generation is not None and uint32(wallet_current_generation) == singleton_record.generation:
+            return
+        to_check: List[SingletonRecord] = []
+        if subscription.mode is DownloadMode.LATEST:
+            to_check = [singleton_record]
+        if subscription.mode is DownloadMode.HISTORY:
+            to_check = await self.wallet_rpc.dl_history(
+                launcher_id=tree_id, min_generation=uint32(wallet_current_generation + 1)
+            )
+        # No root hash changes in the new wallet records, so ignore.
+        if (
+            old_root is not None
+            and old_root.node_hash is not None
+            and to_check[0].root == old_root.node_hash
+            and len(set(record.root for record in to_check)) == 1
+        ):
+            await self.data_store.set_wallet_generation(tree_id, int(singleton_record.generation))
```

Comment:
is this really needed ?
we don't fetch anything in this path so not sure why we update the generation

Reply:
all hashes are identical here - so increase generation so that we return earlier next time: at the condition with checking singleton's generation against our generation.

Reply:
so that is not the real generation we hold ? it a +1 used for earlier exit ?

Reply:
if say the hash stored in data store is {deadbeef, generation=5} but on chain data is initially {deadbeef, generation=5}, then gets updated as  {deadbeef, generation=6},  {deadbeef, generation=7}, we will update here the data store generation to {deadbeef, generation=7}. then, next time it'll detect there's no real change and avoid the extra `dl_history` call (from 5 to 7).

Reply:
thanks for clarifying